[C++] Improve flatbuffers + gRPC integration

llchan opened the pull request:
This PR is a work in progress that stems from some experiments surrounding #4306.

With this PR, flatbuffers can now build messages in gRPC's reference counted `grpc_slice` buffers, allowing us to transfer ownership to gRPC without a copy. This simplifies some of the bookkeeping surrounding buffer ownership and timing, and also reduces the amount of redundant copying (especially when a message is broadcast to many clients).

New API:
- `flatbuffers::BufferRef<T>` becomes `flatbuffers::grpc::Message<T>`, which is a buffer with memory managed by a `grpc_slice`. Most of the reading API remains the same (i.e. `Verify` and `GetRoot`), so other than a type change most existing code should work as-is.
- `flatbuffers::grpc::MessageBuilder` is used instead of `flatbuffers::FlatBufferBuilder` . This is a collection of extra functionality bolted onto `FlatBufferBuilder`:
    - It forces the use of a `flatbuffers::grpc::SliceAllocator`, which does the `grpc_slice` allocation
    - It adds `GetMessage<T>()` and `ReleaseMessage<T>()` methods, which retrieve `Message<T>` objects backed by subslices of the underlying buffer (refcounts are incremented accordingly). The release variant detaches the builder from the message so that the builder can be reused safely without mangling in-flight data.
    - When the building is done and we want to finalize the message, instead of constructing a `BufferRef<T>` from a ptr/size into the buffer, we call `GetMessage<T>()` or `ReleaseMessage<T>()` to obtain a properly reference-counted message that we can then write back as the rpc response.
    - The rest of the `FlatBufferBuilder` methods remain unchanged, so most of the the actual message building code should remain unchanged. 

A few changes to existing code:
- I inherit from FlatBufferBuilder to have access to the underlying buffer while retaining most of the builder methods. To facilitate this, the `final` was removed and `private` was converted to `protected`. There may be a better way to structure this.
- The `vector_downward::release` code was incorrect. The deleter was always using the `simple_allocator::deallocate` instead of the fully derived `deallocate`. This is fixed with a lambda, but depending on compatibility requirements we may need to do something else.
- I exposed the underlying `uint8_t *buf_` pointer of `vector_downward` via a `buf()` method, so we can look up the `grpc_slice` corresponding to the current buffer. If there are objections we could potentially do this by checking which memory region a pointer belongs to, but I don't see any harm in exposing it.

A few warts:
- I originally had the `SliceAllocator` as a member and not a base class of `MessageBuilder`, but the initialization order wasn't right and the `FlatBufferBuilder` would try to use it before it was ready. We can probably refactor something to make this better. Relatedly, there's an extraneous `simple_allocator` member in the FlatBufferBuilder that doesnt need to be there.
- I marked the `SliceAllocator` members as `mutable` because there are some const allocators floating around. Why are they const? I didn't change anything yet but I think we can just pass around non-const allocators, since it's not unreasonable that they might keep state that `allocate`/`deallocate` could mutate.

Anyways, this branch passes the `grpctest` program, but more work needs to be done on the testing front, and we should get more eyes on this to verify the reference counting and such. I'm very much open to feedback and we can iterate on this until everyone's happy with it, so let me know what you think!

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -22,49 +22,198 @@
+class SliceAllocator : public simple_allocator {
+ public:
+  inline SliceAllocator()
+      : slices_{grpc_empty_slice(), grpc_empty_slice()} {
```
Generally, where it really doesn't matter, we like to keep our usage of C++11 as low as possible to support an as wide range of compilers as possible. Our CI includes VS2010 which may not like this new initialization syntax.

llchan replied:
Noted. I'll update these.

ctiller replied:
Note that gRPC does need C++11, so I expect a full CI with VS2010 is going to barf anyway :)

aardappel replied:
Ah yes, good point. I'm realizing we don't have the GRPC code as part of our CI.. which is bad.. but yes, for the moment,.. we're cool with whatever C++11 GRPC uses :)

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -22,49 +22,198 @@
+      slices_[1] = grpc_slice_malloc(size);
+      return GRPC_SLICE_START_PTR(slices_[1]);
+    } else {
+      throw std::logic_error("SliceAllocator full");
```
We may run in environments that have exception handling turned off. Prefer asserts whenever these are programmer errors (as opposed to data errors).

llchan replied:
Got it, I'll assert and return a `nullptr`.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -22,49 +22,198 @@
+
+  template <class T>
+  inline Message<T> GetMessage() const {
+    assert(buf_.data() > buf_.buf());
```
Maybe some comments here (and elsewhere in this file, where mildly tricky things happen). E.g. how does the slice allocator handle the fact that FlatBuffers leaves empty space at the start of a buffer? Doesn't GRPC assume that data starts at the start of a slice?

llchan replied:
Yeah, I can document this inline with the code. Essentially this extracts a subslice of a `grpc_slice` that shares the refcount, so that we don't need to make a copy to reference the message we are sending out.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -22,49 +22,198 @@
+      const flatbuffers::grpc::Message<T> &msg,
+      grpc_byte_buffer **buffer,
+      bool *own_buffer) {
+    *buffer = grpc_raw_byte_buffer_create(const_cast<grpc_slice *>(&msg.BorrowSlice()), 1);
```
some comments as the the exact consequences of this code.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -22,49 +22,198 @@
+  // Deserialize by pulling the
+  static grpc::Status Deserialize(
+      grpc_byte_buffer *buffer,
+      flatbuffers::grpc::Message<T> *msg) {
```
some more comments here too

aardappel commented:
Nice work.. see my comments.
Oh, and not sure why the const on the allocator. That could be removed.

aardappel commented:
Note: this was just merged: https://github.com/google/flatbuffers/pull/4305

llchan commented:
Thanks for the comments. I'll respond inline.

Note that I had some issues with the allocator as it's implemented right now, and sent over another PR #4312 to address those limitations/bugs. I'll rebase this PR once that stuff has been reviewed.

llchan commented on grpc/tests/grpctest.cpp:
```diff
@@ -55,17 +52,16 @@ class ServiceImpl final : public MyGame::Example::MonsterStorage::Service {
 
  private:
-  flatbuffers::FlatBufferBuilder fbb_;
+  flatbuffers::grpc::MessageBuilder fbb_;
```
Yes, I agree that we should beef up documentation on this. I'll contribute what I can with this PR, but we'll probably have to keep expanding over time. I hope to see gRPC + flatbuffers as well documented as the default protobuf usage.

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -576,12 +576,14 @@ class vector_downward {
-    // Point to the desired offset.
-    unique_ptr_t retval(data(), deleter);
+    // Point to the desired offset, but delete from the start of alloc mem
+    unique_ptr_t retval(data(), [&allocator, oldbuf](uint8_t *) {
```
Yep, the ownership is a little incorrect in this PR, actually. Turns out the previous deleter was copying the allocator via copy ctor (notice it was binding to `allocator_` and not `&allocator_`), so the old code was safe. `SliceAllocator` in this PR is not copy-constructible because it's stateful, so the old code wasn't working and I slapped the lambda on it, but that's not really right either.

The other PR I just sent #4312 fixes this for both scenarios. Once that's addressed I'll rebase this PR.

llchan commented on include/flatbuffers/grpc.h:
```diff
@@ -22,49 +22,198 @@
+class Message {
+ public:
+  inline Message()
+      : slice_(grpc_empty_slice()) {
```
Do you know someone who you could pull in to the review this? I'd like to make sure it's correct, and I don't trust my "skim the grpc code for an afternoon" level of understanding.

aardappel replied:
I'll ask someone..

ctiller commented on include/flatbuffers/grpc.h:
```diff
@@ -22,49 +22,198 @@
+  }
+
+  virtual inline uint8_t *allocate(size_t size) const override {
+    if (GRPC_SLICE_IS_EMPTY(slices_[0])) {
```
Is this somehow guaranteed to be called only twice?

llchan replied:
It's only intended to be used by flatbuffer's internal `vector_downward`, and there there is one-to-one correspondence between them (i.e. a single vector owner per allocator) due to the copy ctors being disabled. It needs two because `reallocate` has one in flight while it does the `memcpy`. 

In theory the allocator should only need one `grpc_slice`, and if we are up for changing the implementation a bit, it would simplify things to shift that into a `reallocate` func on the allocator that takes care of the memcpy and such. Let me get some of the other PR's sorted out and I'll update this one.

If we are worried about people seeing/using the allocator without knowing it's specific limitations, we could plop it into a `flatbuffers::grpc::details` namespace or something, or disable the ctor outside of predefined friend classes.

ctiller commented:
LGTM.

llchan commented:
Disregard the many commits above, I rebased off my other PR to get some changes tested. Once that's merged this will be a single-commit PR.

per-gron commented on include/flatbuffers/grpc.h:
```diff
@@ -23,49 +23,233 @@
+    }
+  }
+
+  inline const T* GetRoot() const {
```
I really like that you are exposing only `const` access to `Message`'s underlying buffer. Exposing it mutable has no good use case that I can see and it makes it a lot more scary to pass around and use.

per-gron replied:
Actually, I was reading this wrong. Because of `BorrowSlice` I think you are exposing mutable access to the buffer.

llchan replied:
Yeah, you can get to a mutable buffer via `BorrowSlice()`, so `Message<T>` isn't really a bulletproof immutable object. I'll have to poke around to see if there's a way to make a `grpc_slice` have a const buffer but a non-const refcount. If we're worried about `BorrowSlice()` being exposed, I could probably do something with private + friend class/func where it's used internally, but generally I prefer to let the users shoot themselves in the foot rather than having them frustrated that they can't.

There is perhaps a use case for a non-const accessors, since flatbuffers support mutations (and I have used it in my code before, so it's not just an unused feature). But for something like that I'd probably add a `GetMutableRoot()` func to make it explicit.

per-gron commented on include/flatbuffers/grpc.h:
```diff
@@ -23,49 +23,233 @@
+
+// Message is a typed wrapper around a buffer that manages the underlying
+// `grpc_slice` and also provides flatbuffers-specific helpers such as `Verify`
+// and `GetRoot`.
```
I think it would help to expand this doc comment and say that `Message`, like `grpc_slice`, has shared reference counted ownership of the underlying buffer. This is obvious if you know `grpc_slice` and read the code but it would still help I think.

llchan replied:
Will do, thanks for the suggestion

per-gron commented on include/flatbuffers/grpc.h:
```diff
@@ -23,49 +23,233 @@
+    // Calculate offsets from the buffer start
+    size_t begin = buf_.data() - buf_.buf();
+    size_t end = begin + buf_.size();
+    // Get the slice we are workign with
```
nit: typo "workign"

llchan replied:
Fixed

per-gron commented:
I think this is great! I love that you are taking it all the way to actual zero-copy, that is super useful.

That this is an API break is not a problem for me at all since I'm still just in the experimentation phase of what I want to do with gRPC/Flatbuffers. We both seem to agree that an API break is necessary to make this safe.

From what I can tell, I would not have any issues using this API as is, but if I put on my project maintainer hat on, I would not like that `MessageBuilder` inherits from `FlatBufferBuilder`, and especially not that member variables of `FlatBufferBuilder` are made `protected`. Doing so is [against the Google code style](https://google.github.io/styleguide/cppguide.html#Access_Control) by the way, and that is a rule that I think is probably more important than stuff that clang-format can fix. That these are made `protected` means that almost any change that is done to `FlatBufferBuilder`'s implementation risks becoming an API break. If you must keep the inheritance, I would prefer to make `FlatBufferBuilder` and `MessageBuilder` `friend`s instead, to limit API break exposure.

Personally, I would like more if `FlatBufferBuilder` would be made a template that takes the allocator type as a template parameter (with a default value or a different name for the template class + typedef to not break the normal `FlatBufferBuilder` use case) (this would make the `virtual` methods on the allocator redundant but that's not the point), because then the allocator type could expose something that has the bulk of `MessageBuilder::GetMessage`'s functionality. Then, `FlatBufferBuilder` could have a method that calls the allocator to do `MessageBuilder::GetMessage<T>` does here, and `MessageBuilder` could be a typedef of `FlatBufferBuilder<SliceAllocator>`. Then there would be no need for protected member variables or inheritance.

I haven't thought this through super carefully so it might not work out but I think it could be worth spending a little bit of time trying to refine it. I have a feeling that current inheritance + protected members scheme could become increasingly messy over time.

llchan commented:
Regarding the comment about inheriting from `FlatBufferBuilder`: I generally do agree with the composition over inheritance mantra, but in this case I wanted the existing generated code to work with it without changes, and the generated builders operate on `FlatBufferBuilder` references. I also didn't want to have to forward every method, instead letting the inheritance take care of that.

You also mentioned templates, and in fact my original implementation did exactly that. I templated `vector_downward` and `FlatBufferBuilder`, very similar to what you proposed. However, @aardappel felt like this would be too large of a change, so as requested I modified the changes to rely on inheritance instead.

aardappel commented:
I agree that inheriting is not the nicest way of doing things, but if this gets us the simplest/cleanest code, I'm ok with it. Certainly forwarding all methods is a mess I'd not be happy with. I guess the alternative would be for `MessageBuilder` to own a `FlatBufferBuilder`, and once it has set up the correct allocator, allows the user to retrieve a reference to the `FlatBufferBuilder`. Any disadvantage in that compared to inheritance?

Yes, my fault for pushing to not make `FlatBufferBuilder` into a template. Maybe I am being too conservative, and if others agree that templates would bring strong benefits, I can be overruled :) .. we only just changed the API, after all.

llchan commented:
It's always a tradeoff with templates, and I understand the desire for simplicity. I don't feel strongly enough to push for it, especially since it's not a blocking issue and the inheritance approach is also working (albeit with slightly less type enforcing at compile-time).

I've rebased this branch off master and updated things as mentioned above.

llchan commented:
I've also added a `vector_downward` buffer accessor so that we don't need `FlatBufferBuilder` members to be protected for `MessageBuilder` to see them. This addresses @per-gron's comment about the added maintenance cost of protected members.

per-gron commented:
Thanks for addressing some of my comments :-) Just to it clear: I wrote some review comments with some of the things I thought about when I read the PR, but I don't at all want that to block merging this PR. I think this is a great improvement and I wouldn't have any problems to use the library if it were like this. Your work is very much appreciated @llchan and @aardappel.

per-gron commented:
@llchan Another thing that might be interesting to think about:

To me, it would be very useful if flatbuffer validation was performed automatically for me, by the `SerializationTraits`. I think not validating input that is likely from network is not a good default setting. Would this be possible? It looks like it, given the return type of `Serialize` and `Deserialize` I guess this could be done in a separate PR though.

llchan commented:
@per-gron yeah that might be handy to save the user a check in every rpc call. How would it look? Would we want `flatbuffers::grpc` to have its own serialization traits? We could do something like this in the `flatbuffers::grpc` namespace, though maybe a different name might be better:

```c++
// default is to verify
template <class T>
class SerializationTraits<T> {
  static constexpr bool verify = true;
};

// user can turn it off if they know they are dealing with trusted sources
template <>
class SerializationTraits<Monster> {
  static constexpr bool verify = false;
};
```

llchan commented:
Actually maybe something better would be something like this:

```c++
template <class T>
class MessageVerifier {
  static inline bool Verify(const Message<T>& msg) const {
    // the usual verify code
  }
};
```

The user can then choose to stub it out with a no-op, they could use the default verifier, and they can even add additional checks (e.g. if fields have constraints not expressed in the flatbuffers schema).

llchan commented:
@per-gron I've added message-verification-by-default in the latest commit. There is a fundamental limitation with the current traits-based toggling: for a given type, you can only really specify a single type of verification behavior. That is, you can't have two different rpc methods accepting a `Monster`, and have one verify the message but not the other. This is not really a show-stopper, since it should be possible to work around this with some wrapper types, but just something to be aware of.

I'm open to other implementation ideas if you have any.

per-gron commented:
@llchan Thanks for adding verification, that will be very helpful for me.

I like the approach that you came up with that allows further type-specific verification, that sounds like a useful additional feature (although the user will have to be careful about actually including that template specialization always).

per-gron commented:
A somewhat safer approach could be to allow the user to specify if validation should happen or not (or custom via trait) in the .fbs file. But that seems a bit over-engineery to me. Just doing verification by default seems sane to me and anything else is ok if it's a bit tricky to get at imo.

llchan commented:
Interesting idea to put it in the schema for safer guarantees, but I'm not too familiar with that part of the code so I'll defer to another contributor/maintainer.

aardappel commented:
@per-gron : Not sure I agree the verifier should be default on. While typically you'd want to use it for untrusted network traffic, it should still be a conscious decision to use it, and what you'd want to do if it fails will differ by the use-case. People also use GRPC between trusted servers in the same internal network / data center, and may not want to pay the (small) cost of the verifier when maximum thru-put is the goal.

We could add it to the examples/comments though, indicating people should consider using it.

llchan commented:
I don't have a preference regarding the default verification behavior, since it's easy to override with `MessageVerifier<T>` specializations.

The latest PR disables it by default, adds a preprocessor flag that can enable it by default, and adds a helper function to `Message<T>` so specializations can also enable it on a type-by-type basis.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -697,6 +697,24 @@ class vector_downward {
   }
 
+  void reset(size_t initial_size = 1024, Allocator *allocator = nullptr,
+             bool own_allocator = false) {
```
Why not simplify this method to assume we keep the same allocator, since it is only used from GRPC code at the moment? In fact, it is called with the exact same parameters than on construction, so I don't see why we're reallocating the buffer.

llchan replied:
I'm taking a look at this now, it might be a vestige of some older impl details with an older iteration of the allocator API.

llchan replied:
Okay I took a look at this and it was indeed an artifact of some older code. It wasn't performing a dynamic allocation, but it was setting some allocator pointers that don't need to be set (they don't change).

I've updated this to just be a plain `reset()` func that deallocates the old buffer and reallocates a new one with a capacity based on the original `initial_size`.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -853,6 +882,16 @@ FLATBUFFERS_FINAL_CLASS
 
+  /// @brief Get the underlying `vector_downward`
+  /// @return Returns the underlying `vector_downward` buffer
+  vector_downward &GetVectorDownward() { return buf_; }
```
If we're going to be using inheritance, neither of these methods are necessary for access, and add unnecessary public functionality

llchan replied:
This was added primarily as a response to @per-gron concern that protected inheritance can lead to implementation coupling that isn't necessarily obvious from the public interface. I don't have a preference either way, so I'll change this to the original protected access mode change.

aardappel replied:
I have no problem with `protected`.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,271 @@
+ public:
+  inline Message() : slice_(grpc_empty_slice()) {}
+
+  inline Message(grpc_slice slice, AddRef) : slice_(grpc_slice_ref(slice)) {}
```
This is a neat pattern, but since this is mostly used internally, I think a bool would be sufficient too.

llchan replied:
I based this on a `grpc_slice` wrapper I saw in the gRPC code. One benefit that I could see is that it makes the calls more readable (i.e. you see `ADD_REF` rather than `true`), but could also be done with a single enum type. I suppose making it two distinct types means they can have to completely different impls for the two ctors.

In any case, the ctors here are trivial so I'm happy to make it a bool.

aardappel replied:
It's both more readable and also less, since it took me a bit of time to figure out why the types were needed :)

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,271 @@
+  }
+
+  inline const uint8_t *data() const {
+    if (slice_.refcount != nullptr) {
```
Hmm, what a clumsy API. Maybe maybe one function that returns either `slice_.data.refcounted` or `slice_.data.inlined` and use it twice?

llchan replied:
Unfortunately the structs in the union have different layouts (i.e. length is at different places). I'm guessing they did this deliberately either for optimization or to prevent people from making the assumption that they can rely on a consistent layout.

Or, maybe I'm misunderstanding your suggestion?

ctiller replied:
why not just use the GRPC_SLICE_START_PTR(slice_) macro?

llchan replied:
Probably because I wrote the message wrapper before I discovered those macros :)

aardappel replied:
Ah, thanks @ctiller :)

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,271 @@
+    return verifier.VerifyBuffer<T>(nullptr);
+  }
+
+  inline ::grpc::Status VerifyGRPC() const {
```
nice!

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,271 @@
+
+// MessageBuilder is a gRPC-specific FlatBufferBuilder that uses SliceAllocator
+// to allocate gRPC buffers.
+class MessageBuilder : private detail::SliceAllocatorMember,
```
rather than this hack, what would be the downside of making both base classes members, and adding a `GetFlatBufferBuilder()` ? is calling such an extra function unwelcome verbosity? :)

llchan replied:
I'd personally prefer to avoid the extra verbosity, because it will just be an extra line of boilerplate in every single RPC func. No one really sees this implementation detail, so maybe it's not so bad?

aardappel replied:
Ok, I guess so.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,271 @@
+  MessageBuilder(const MessageBuilder &other) = delete;
+  MessageBuilder &operator=(const MessageBuilder &other) = delete;
+
+  inline ~MessageBuilder() {}
```
I don't think we need `inline` in front of inline methods, they're already guaranteed inline.

llchan replied:
Sure, I can remove them, just a habit of mine to spell it out when it's in a header :)

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,271 @@
+  template <class T>
+  inline Message<T> GetMessage() {
+    vector_downward &vec = GetVectorDownward();
+    uint8_t *buf_data = vec.buf();     // pointer to memory
```
use `auto` where possible.

llchan replied:
Is there a rule for when to use auto? Everything inside a function definition?

aardappel replied:
There's no such rule in the Google style guide, but the FlatBuffers projects uses it whereever it is valid.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,271 @@
+
+    // Get the slice we are working with (no refcount change)
+    grpc_slice slice = slice_allocator_.get_slice(buf_data, buf_size);
+
```
Google style: no need for all these empty lines between each statement

llchan replied:
Can do. Was only included for readability.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,271 @@
+  template <class T>
+  inline Message<T> ReleaseMessage() {
+    Message<T> msg = GetMessage<T>();
+    Reset(initial_size_, &slice_allocator_, false);
```
Like I said, even if `deallocate` is just an unref, there's still an allocation in `reset` that is not needed, and will just get deallocated shortly after if this `MessageBuilder` is abandoned.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,271 @@
-                   reinterpret_cast<const char *>(msg.buf), msg.len);
-    *buffer = grpc_raw_byte_buffer_create(&slice, 1);
-    grpc_slice_unref(slice);
+  static inline ::grpc::Status Verify(const Message<T> &msg) {
```
What is this class needed for? Can we inline this code?

aardappel replied:
Ah ok, so it is overrideable. So now we have both a pre-processor flag and an overrideable template, just to save the user from calling a function manually.. this may be a bit much. If we have `Verify` in all examples and documentation, that should be enough for users to consider if they want it.

llchan replied:
It may be a bit overkill, but if someone wants to be safe, the preprocessor flag is a much better guarantee than making sure they specialize every type (not to mention less code).  I agree we will want this documented.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,271 @@
-    grpc_byte_buffer_reader_destroy(&reader);
-    grpc_byte_buffer_destroy(buffer);
-    return grpc::Status();
+    return flatbuffers::grpc::MessageVerifier<T>::Verify(*msg);
```
Ok, I now see that this is kind of nice, since you now have one point of error checking when you receive the message. Once concern is though that if you check for status and get a `::grpc::StatusCode::DATA_LOSS`, you must be able to know that this is a verifier failure, not some generic networking error. So we must clearly document that a) you can set `FLATBUFFERS_GRPC_AUTO_VERIFY`, and b) that if you do, you should listen for that code. if `DATA_LOSS` is used also by other GRPC conditions, there must be a way to find out if it was a verifier failure without running the verifier again. Storing a bool in `Message` doesn't seem that elegant.

When you add it all up, it seems almost nice to just suggest to the user they `Verify` on the message themselves when they receive it, what do you think?

llchan replied:
Yeah, I wasn't totally certain what code is appropriate for a deserialization error, if you could ask gRPC folks that might be helpful to get the best practice code to return.

I don't think the gRPC server needs to do anything with the error code. I'm not positive on this, but I think when gRPC goes to deserialize and detects an error return, it won't even call the RPC func at all, so the user shouldn't need to worry about it.

Client code might have to handle it, but if they're sending over malformed messages there's something wrong anyways.

aardappel replied:
@ctiller is `::grpc::StatusCode::DATA_LOSS` appropriate for when the serializer finds corrupt data? Or is there a better way to handle this?

llchan replied:
Maybe it could also be `INVALID_ARGUMENT`. When I read through the [descriptions](http://www.grpc.io/grpc/cpp/namespacegrpc.html#aff1730578c90160528f6a8d67ef5c43b) originally, I interpreted that code as a way for the rpc call itself to indicate that one of the arguments inside the request was invalid, not that the request was corrupted/unparseable.

`DATA_LOSS`'s description was "Unrecoverable data loss or corruption.", so I originally used that.

I'm fine with any suggestion, just lmk.

llchan replied:
Thanks for looking that up, I agree we should try to behave like protobuf, so unless there's a strong opinion otherwise I'll update to use that.

aardappel replied:
agreed its good to behave the same. still would be nice to be able to know if its a verifier failure or something else, though.

aardappel commented:
@per-gron I think that depends on your use case.

llchan commented:
Based on the discussion in the issue and comments here, I _think_ the consensus is that the gRPC verifier should at least run the basic verification by default to ensure that message access will be safe.

I'd like to do docs in separate PR since this one is already getting fairly lengthy, but I will def include this in the docs.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -697,6 +698,17 @@ class vector_downward {
+    if (buf_ != nullptr) {
+      allocator_->deallocate(buf_, reserved_);
+    }
+    reserved_ = (initial_size_ + sizeof(largest_scalar_t) - 1) &
```
this logic is copied from just above here, maybe add a `roundup()` function?

llchan replied:
Can do. Static inline func sound good?

aardappel replied:
yes.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -697,6 +698,17 @@ class vector_downward {
+    }
+    reserved_ = (initial_size_ + sizeof(largest_scalar_t) - 1) &
+                ~(sizeof(largest_scalar_t) - 1);
+    buf_ = allocator_->allocate(reserved_);
```
There is still an allocation here.. As I understand it, in the single use case for this function, we reset this buffer and then throw it away.

llchan replied:
While I'm looking at this, is there a reason for `clear()` or the ctor to allocate? Could we also make that lazy and rely on `make_space(...)` for all allocations? That way an unused or cleared fbb doesnt cost any heap allocations until it's used.

llchan replied:
I should be more precise: `clear()` when there is a nullptr buffer reallocates, and I think we can remove that and let `make_space` handle the reallocation. If the buffer already exists, we can leave it alone and reset the `cur_` pointer.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,247 @@
+  Message(grpc_slice slice, bool add_ref)
+    : slice_(add_ref ? grpc_slice_ref(slice) : slice) {}
+
+#if FLATBUFFERS_GRPC_ENABLE_MESSAGE_COPY_CTOR
```
Should these two defines be merged? copy constructor and assignment typically go together.
Do we need them at all? If they just cause a ref increase, is there any harm in allowing it?

llchan replied:
None of the code as written should be utilizing those, and I disabled them to audit my own code and make sure we're not needlessly modifying the refcount. I think there's some memory fences or other atomics around those for thread safety, so it's not totally free to modify the refcount.

I don't think a general use case should be making copies of the messages, and I'd like for it to be an conscious decision to enable the opt-in feature.

llchan replied:
As for merging the two defines, I'm okay with that, doesnt matter to me.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,247 @@
+    assert(msg_data >= msg_data);
+    assert(msg_data + msg_size <= buf_data + buf_size);
+    // Calculate offsets from the buffer start
+    size_t begin = msg_data - buf_data;
```
auto

llchan replied:
roger

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,247 @@
+};
+
+template <class T>
+class MessageVerifier {
```
Sounds like we could just inline this class, since if you're not going to be using the verifier, there really isn't anything else to replace it with, so no reason to override this. The define takes care of all.

llchan replied:
Do we care about allowing per-type specialization? I left that as an option for flexibility, but maybe that's an unlikely use case, and we can make the assumption that each translation unit either wants verification globally or not.

aardappel replied:
I think we can make that assumption for now. Prefer to usually not make abstractions that have 0 users.. easy to add later if needed.

aardappel commented on grpc/tests/grpctest.cpp:
```diff
@@ -108,15 +104,14 @@ int main(int /*argc*/, const char * /*argv*/[]) {
+    flatbuffers::grpc::Message<Stat> response;
 
     // The actual RPC.
     auto status = stub->Store(&context, request, &response);
```
We should check the status here specifically for verifier failure (and only verifier failure, not other occurrences of DATA_LOSS).

llchan replied:
Unless there's specifically a code added for "parsing error" or somesuch, I think the best we can do is check the error string, which might be a little brittle. But I can certainly add that.

Btw, in case you didn't see, I'll be updating the error code to `INTERNAL` to match the protobuf gRPC behavior when the message is unparseable.

llchan commented:
I've made the requested changes, currently working on merging some `base.h` changes that were introduced in master.

llchan commented:
Okay merge is done. Most of the changes in the last iteration were fairly minor/trivial. The only one that changes behavior materially is the change to make `vector_downward` uniformly lazy for all code paths: `ctor`, `clear`, and `reset`. It will allocate on the first write i.e. first `make_space(..)` call, making builders fully allocation-free to instantiate.

aardappel commented on grpc/tests/grpctest.cpp:
```diff
@@ -133,18 +128,26 @@ int main(int /*argc*/, const char * /*argv*/[]) {
+    flatbuffers::grpc::Message<Stat> response;
+    auto status = stub->Store(&context, request, &response);
+    assert(!status.ok());
+    assert(status.error_code() == ::grpc::StatusCode::INTERNAL);
```
maybe some explanation comments here?

llchan replied:
sure thing

aardappel commented on include/flatbuffers/base.h:
```diff
@@ -165,6 +165,10 @@ template<typename T> void WriteScalar(void *p, T t) {
   *reinterpret_cast<T *>(p) = EndianScalar(t);
 }
 
+static inline size_t RoundUp(size_t size, size_t round_size) {
```
base.h is shared between FlatBuffers and FlexBuffers. If this is not used by FlexBuffers it doesn't need to be here. In fact, I was expecting it to be just a static class method since it is only used inside one class.

aardappel replied:
Actually it appears to be only used in one place now, so it can be inlined again :)

llchan replied:
yep, i'll put it back inline

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -437,14 +437,14 @@ class DetachedBuffer {
     : allocator_(allocator), own_allocator_(own_allocator), buf_(buf),
       reserved_(reserved), cur_(cur), size_(sz) {
-    assert(allocator_);
+    assert(allocator_ != nullptr);
```
The style thru-out the FlatBuffers code is to check pointers without `!= nullptr`. So rather than adding it in a few more places, it would be more consistent to remove it everywhere.

llchan replied:
fine by me, making this change

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -616,19 +640,29 @@ class vector_downward {
-    buf_ = allocator_->reallocate_downward(buf_, old_reserved, reserved_);
-    cur_ = buf_ + reserved_ - old_size;
+    size_t reserved;
+    if (old_reserved == 0) {
```
use `?:` ?

aardappel replied:
auto reserved = old_reserved + (std::max)(len, old_reserved == 0 ? initial_size_ : growth_policy(old_reserved));
And why the cast? make initial_size_ a size_t?
Why both `reserved` and `reserved_`? harder to follow.
In fact, since we can already tell from`buf_` there is no buffer, we don't need `reserved_` to signal there's no buffer, and could have kept the older, simpler logic.

llchan replied:
Actually now that I look at this, we should just put the `reserved_ == 0` case in `growth_policy(...)`.

`initial_size` is a `uoffset_t` because that's how it was in the original implementation. I think it could be a `size_t` across the board, so I'll go ahead and do that if we don't care about the change in size/alignment.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,232 @@
+  Message(grpc_slice slice, bool add_ref)
+    : slice_(add_ref ? grpc_slice_ref(slice) : slice) {}
+
+#if FLATBUFFERS_GRPC_ENABLE_MESSAGE_COPY
```
as I asked before, why is this necessary?

llchan replied:
Might have missed my previous response, pasting i below:

None of the code as written should be utilizing those, and I disabled them to audit my own code and make sure we're not needlessly modifying the refcount. I think there are some memory fences or other atomics around those for thread safety, so it's not totally free to modify the refcount.

I don't think a general use case should be making copies of the messages, and I'd like for it to be an conscious decision to enable the opt-in feature.

aardappel replied:
ok!
I guess in the future making comment in code as to why you probably don't want to make copies is helpful.

llchan replied:
Actually if your point is that we should remove them entirely, I'm okay with that.

I don't know for sure if gRPC buffers are reused. If they are, then messages are not safe to keep outside of the rpc function, and there pretty much be no reason to copy/assign a `flatbuffers::grpc::Message<T>`. @ctiller, could you comment on the gRPC buffer reuse policy just so I know?

I think in either case I'll just remove these, since the semantics of a copy might be misleading if it's still tied to the same buffer.

aardappel commented on include/flatbuffers/grpc.h:
```diff
@@ -23,50 +23,232 @@
-    grpc_byte_buffer_reader_destroy(&reader);
-    grpc_byte_buffer_destroy(buffer);
-    return grpc::Status();
+#if FLATBUFFERS_GRPC_DISABLE_AUTO_VERIFICATION
```
Indent pre-processor statements. I know that's odd, but that's what the rest of the code does.

llchan replied:
So clang-format wants to put them back at column 0... is there a flag I need to add? More generally, could we include a `.clang-format` so I don't need to think about these things?

llchan commented:
I added some example code + simple docs (and fixed a builder reuse bug I found while doing so). My priority is to get this PR merged in, so if the docs are not up to par and you don't want them on master, I can rip them out and we can deal with them in a separate PR.

Just thought I'd include a first pass at it since people sometimes like changes to be accompanied with the corresponding doc changes in the same PR.

aardappel commented on docs/source/gRPC/CppUsage.md:
```diff
@@ -0,0 +1,29 @@
+Use in C++    {#flatbuffers_grpc_guide_use_cpp}
+==========
+
+## Before you get started
```
Thanks! Docs are better than no docs, so we can leave this as-is for this PR.

aardappel commented on examples/grpc/greeter/Makefile:
```diff
@@ -0,0 +1,14 @@
+CXXFLAGS ?= -I../../../include
```
Since we want this to be portable, eventually this should become a CMakefile.txt.. but we can leave it for now if you want.

llchan replied:
Yeah, I don't have much experience with cmake, and I don't have non-Linux platforms to test on, so I don't feel qualified to write the `CMakeFile.txt` for the example. I'll leave it as-is for now and someone else can update it later.

aardappel commented on examples/grpc/greeter/client.cpp:
```diff
@@ -0,0 +1,84 @@
+#include "greeter.grpc.fb.h"
```
Thanks for adding another example!
Not sure about the top level `examples` dir, since we already have `samples` we could put it under, or better yet use `grpc/samples` since we already are keeping everything grpc related under there, including tests.

llchan replied:
Not sure why I overlooked the `samples` directory, maybe just a reflex since `examples` is more common. I'll move them into `grpc/samples`.

aardappel commented on examples/grpc/greeter/client.cpp:
```diff
@@ -0,0 +1,84 @@
+
+    auto status = stub_->SayHello(&context, request_msg, &response_msg);
+    if (status.ok()) {
+      const HelloReply* response = response_msg.GetRoot();
```
style: location of `*`, or better yet, auto.

llchan replied:
Yep, this was a clang-format misconfiguration, fixed.

aardappel commented on examples/grpc/greeter/client.cpp:
```diff
@@ -0,0 +1,84 @@
+    }
+  }
+
+  template <class F>
```
style, `template` on same line as the rest.

llchan replied:
I saw a mixture of inlined and separate-lined templates in `flatbuffers.h` so I thought it was a "whichever fits better on the line" rule, but now that I look I think it's a "inline for methods and separate-line for regular funcs". Is that correct?

Is there any way to express that rule in clang-format? Or do I need to manually format these? Or do you guys use a different formatter? I'm trying my best to infer conventions, but requiring contributors to get formatting details right without automated tooling is kind of asking a lot.

llchan replied:
Well, this is somewhat irrelevant now that I've switched it to a `std::function` based on the comment below. But I still think we should provide formatting configs if we expect very specific formatting.

aardappel replied:
It's not a big deal either way, there's certainly some inconsistencies in style due to external contributors, things I didn't catch.

No idea how to express that in clang-format. There's quite a few things the Google C++ Style Guide leaves unspecified, so it defaults to "how the project has always done it".

I'm a big fan of clang-format, but running it will typically polute your commit with things from older changes. Unless we enforced running it on merge somehow it's hard to rely on it. It also doesn't support the pre-processor indentation and a few other things. Usefully, some editor integrations allow you to clang-format just a selection.

aardappel commented on examples/grpc/greeter/client.cpp:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  template <class F>
+  void SayManyHellos(const std::string& name, int num_greetings, F&& callback) {
```
if its not performance sensitive, maybe use `std::function` for callbacks? or maybe no callback at all, since in an example, simple straightline code to follow is more important than modularity/factoring concerns.

llchan replied:
Sure, it doesnt matter to me. I mainly wanted it to be symmetric with the `SayHello` code, where the "thing to do" was defined in the main func and not in the `GreeterClient`.

llchan replied:
Btw, this example is based off the gRPC cpp example.

aardappel commented on examples/grpc/greeter/client.cpp:
```diff
@@ -0,0 +1,84 @@
+
+ private:
+  std::unique_ptr<Greeter::Stub> stub_;
+  flatbuffers::grpc::MessageBuilder mb_;
```
if this is an instance variable because you want to reuse it, you need to be calling `Reset` on it.

aardappel replied:
I mean `Clear` :)

llchan replied:
The `ReleaseMessage<T>()` function actually detaches the message/buffer from the builder, so the builder is already cleared after each request is sent out. I can make it a local var if you prefer, I only did it this way because the grpctest implementation kept the builder in a member var.

aardappel commented on examples/grpc/greeter/server.cpp:
```diff
@@ -0,0 +1,80 @@
+    const HelloRequest *request = request_msg->GetRoot();
+
+    // Fields are retrieved as usual with FlatBuffers
+    const std::string &name = request->name()->str();
```
`auto &`, here and elsewhere.

llchan replied:
Yep, this was a clang-format misconfiguration. Fixed.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -69,7 +69,7 @@ template<typename T> T EndianSwap(T t) {
 }
 
-template<typename T> size_t AlignOf() {
+template<typename T> constexpr size_t AlignOf() {
```
what is this needed for? there's no `constexpr` anywhere else in the code, so I'd rather not have it for the sake of older compilers, for the moment.

llchan replied:
This is used in the reallocate func, where the alignment is marked as a constexpr. I saw that there were already preprocessor checks for a constexpr fallback to const, so I thought we'd already handled portability for older compilers?

aardappel replied:
you're right, we do.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -616,18 +638,23 @@ class vector_downward {
-    reserved_ = (reserved_ + (largest_align - 1)) & ~(largest_align - 1);
-    buf_ = allocator_->reallocate_downward(buf_, old_reserved, reserved_);
+    reserved_ += (std::max)(len, growth_policy(old_reserved));
+    constexpr size_t alignment = AlignOf<largest_scalar_t>();
```
remove `constexpr` if possible.

llchan replied:
Sure, but what's the harm? We already did the preprocessor dance to make it work with older compilers?

aardappel replied:
yeah, nevermind.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -685,6 +709,16 @@ FLATBUFFERS_FINAL_CLASS
     if (string_pool) delete string_pool;
   }
 
+  void Reset() {
```
This appears to be a duplicate of `Clear` below, so can be removed entirely.

llchan replied:
It's not entirely the same when there's an existing buffer. `Clear` reuses, and `Reset` deallocates (with a lazy allocate when it's first used next). This is an important distinction with the gRPC slices, where it's not safe to reuse the buffers.

aardappel replied:
Can one reuse the other, or can code be shared?

llchan replied:
We could reset the vector after we clear, which would have the same end result. Un-optimized, there'd be an extra store to the `cur_` pointer, but I _think_ the compiler should be able to eliminate that first store (correct me if I'm wrong). I'll update to do so, since either way the cost is minimal/zero.

aardappel commented:
Ok, this all looks good to me! Any additional improvements lets add them in follow-up PRs.
Thanks for the hard work :)

cberner commented:
This looks great! What version of gRPC do I need to use this with? Just tried compiling a simple service, but end up with a compiler error:
```
parameter_server_fb.grpc.fb.cc:38:10: error: 'Create' is not a member of 'grpc::ClientAsyncResponseReader<flatbuffers::grpc::Message<PS::Parameters> >'
   return ::grpc::ClientAsyncResponseReader< flatbuffers::grpc::Message<Parameters>>::Create(channel_.get(), cq, rpcmethod_SyncParameters_, context, request);
```

llchan commented:
Yep, it needs master or the 1.4.0 pre-release. For what it's worth, that requirement was already introduced into master before this PR, but doesn't hurt to mention it here too.

gwvo commented:
@llchan @per-gron you may be interested in this up-coming blog post, which is based on your work: https://github.com/grpc/grpc.github.io/pull/550

per-gron commented:
Nice!
